itachallenge-auth/src/main/java/com/itachallenge/auth/controller/AuthController.java

```
@@ -63,29 +67,47 @@ public String test() {
         return "Hello from ITA ChallengeAuth!!!";
     }
 
-
     @PostMapping("/github/authenticate")
-    public Mono<ResponseEntity<Map<String, Object>>> authenticateWithGithub(@RequestBody Map<String, String> codeRequest) {
-        return authService.exchangeCodeForToken(codeRequest.get("code"))
+    public Mono<ResponseEntity<Map<String, Object>>> authenticateWithGithub(@RequestBody(required = false) Map<String, String> codeRequest) {
+        if (codeRequest == null || !codeRequest.containsKey("code") || !codeRequest.containsKey("redirect_uri")) {
+            Map<String, Object> body = new HashMap<>();
+            body.put("message", "Missing 'code' or 'redirectUri' in the request");
+            body.put("isValid", false);
+            body.put("username", null);
+
+            return Mono.just(ResponseEntity.badRequest().body(body));
+        }
+        String code = codeRequest.get("code");
+        String redirectUri = codeRequest.get("redirect_uri");
+
+        if (code == null || redirectUri == null) {
+            return Mono.just(ResponseEntity.badRequest().body(Map.of(
+                    "message", "Missing 'code' or 'redirectUri' in the request",
+                    "isValid", false,
+                    "username", null
+            )));
+        }
+
+        return authService.exchangeCodeForToken(code, redirectUri)
                 .flatMap(authService::validateTokenWithGithub)
                 .flatMap(response -> {
-                    if (!(boolean) response.get(KEY_IS_VALID)) {
+                    if (!(boolean) response.get("isValid")) {
```

```
.flatMap(response -> {
    if (!(boolean) response.get("isValid")) {
        return Mono.just(ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .header("X-Authentication-Status", "Failed")
                .body(response));
    }
    String githubUsername = (String) response.get("username");
    return getUserDetailsFromGithubUsername(response, githubUsername);
})

```
Consider moving this logic into AuthService, and instead of returning the ResponseEntity directly from the controller, throw a custom exception when isValid is false:

---

Solved -- simplified the logic inside the controller.